### DISCUSSION TITLE
How to downcast `connection_limits::Exceeded` ?

### DISCUSSION THREAD

**Discussion by nathanielc** [libp2p/rust-libp2p#4018] [Q&A] [answered]
I am trying to upgrade to libp2p 0.51.0, specifically to use the new `connection_limits::Behaviour` however I am unable to downcast the error cause as explained in the docs. 

> The [ListenError::Denied](https://docs.rs/libp2p/latest/libp2p/swarm/enum.ListenError.html#variant.Denied) and respectively the [DialError::Denied](https://docs.rs/libp2p/latest/libp2p/swarm/enum.DialError.html#variant.Denied) variant contain a [ConnectionDenied](https://docs.rs/libp2p/latest/libp2p/swarm/struct.ConnectionDenied.html) type that can be downcast to [Exceeded](https://docs.rs/libp2p/latest/libp2p/connection_limits/struct.Exceeded.html) error if (and only if) this behaviour denied the connection.

Here is my code snippet

```rust
// Inside impl NetworkBehaviour ...

fn on_swarm_event(&mut self, event: libp2p::swarm::FromSwarm<Self::ConnectionHandler>) {
    match event {
        // ...
        libp2p::swarm::FromSwarm::DialFailure(event) => {
            if let DialError::Denied { cause } = event.error {
                if let Ok(connection_limits::Exceeded { .. }) = cause.downcast() {
                    // Do something about the connection limit error
                }
            }
        } 
        // ...
    }
}
```

I am unable to downcast as I only have a [reference](https://github.com/libp2p/rust-libp2p/blob/master/swarm/src/behaviour.rs#L479) to the `cause`. Downcast needs to move the `cause` value but its behind a reference. I can't clone the value either as it does not implement `Clone`. Am I missing something? 

Happy to contribute a fix if one is needed.

**Comment by thomaseizinger** [comment]
Ah yes, the downcast was intended to be used when you receive the event via polling the `Swarm`.

We can't give ownership of the error to all behaviours because errors are seldomly clonable.

**Comment by mxinden** [comment]
@nathanielc can you post your `NetworkBehaviour` hierarchy here? Are you wrapping the `connection_limits::Behaviour`?

**Comment by nathanielc** [comment]
Yeah makes sense why Clone isn't available. 

Here is the WIP branch I have for more context https://github.com/nathanielc/beetle/blob/custom-behaviour/iroh-bitswap/src/lib.rs#L426

This is in a stripped down fork of beetle specifically in the bitswap implementation. 

Here is the behaviour itself https://github.com/nathanielc/beetle/blob/custom-behaviour/iroh-p2p/src/behaviour.rs#L37-L52 Its mostly flat with exception of the `custom_behavior` that allows us to layer in more behaviours below the beetle behaviours (i.e. https://github.com/3box/rust-ceramic/pull/52)

**Reply by thomaseizinger** [reply]
There is not reason we cannot also expose `downcast_ref`. Mind opening a PR?

**Reply by nathanielc** [reply, marked as answer]
That was simple, let me know if you'd like a different/better test https://github.com/libp2p/rust-libp2p/pull/4020
